R-package/tests/testthat/test_basic.R

```diff
@@ -2207,6 +2208,96 @@ test_that(paste0("lgb.train() gives same results when using interaction_constrai
 
 })
 
+test_that(paste0("Category encoding for R package works"), {
+  # test category_encoders
+  set.seed(1L)
+  dtrain <- lgb.Dataset(train$data, label = train$label)
+  dtest <- lgb.Dataset(test$data, label = test$label, reference = dtrain)
+  cat_fid <- c(1L, 2L, 3L, 4L)
+  # ``` category_encoders = "" ```   is equal to   ``` category_encoders = "raw" ```
+  params <- list(objective = "binary", categorical_feature = cat_fid, category_encoders = "")
+  bst <- lightgbm(
+    data = dtrain
+    , params = params
+    , nrounds = 10L
+    , verbose = 2L
+    , valids = list("valid1" = dtest)
+  )
+  pred1 <- bst$predict(test$data)
+
+  # treat the first 4 features as categorical features
+  dtrain <- lgb.Dataset(train$data, label = train$label,
+    categorical_feature = cat_fid, category_encoders = "raw")
+  dtest <- lgb.Dataset(test$data, label = test$label,
+    categorical_feature = cat_fid, reference = dtrain)
+  params <- list(objective = "binary")
+  bst <- lightgbm(
+    data = dtrain
+    , params = params
+    , nrounds = 10L
+    , verbose = 2L
+    , valids = list("valid1" = dtest)
+  )
+  pred2 <- bst$predict(test$data)
+  expect_equal(pred1, pred2)
+
+  dtrain <- lgb.Dataset(train$data, label = train$label,
+    categorical_feature = cat_fid)
+  dtest <- lgb.Dataset(test$data, label = test$label,
+    categorical_feature = cat_fid, reference = dtrain)
+  params <- list(objective = "binary", category_encoders = "target,count,raw")
+  bst <- lightgbm(
+    data = dtrain
+    , params = params
+    , nrounds = 10L
+    , verbose = 2L
+    , valids = list("valid1" = dtest)
+  )
+  pred3 <- bst$predict(test$data)
+  expect_equal(dim(dtrain), c(6513L, 134L))
```

Comment:
am I right here the the purpose of this test on `dim()` is to check that additional features were added based on the values passed to `category_encoders`?

The `lightgbm::agaricus` dataset has 6513 rows and 126 features, so I guess this is saying "with `target,count,raw`", LightGBM should add 8 new features...one `count` feature and one `target`  feature per categorical feature.

If I'm right about that, then please consider this change to the tests to make it more obvious that that is what is happening here. Otherwise, this might fail some day and we won't know what the hard-coded numbers 6513 and  134 were supposed to mean.

```suggestion
  # one new "count" and "target" feature is added per categorical feature
  num_new_cat_features <- length(cat_fid) * 2L
  expect_equal(dim(dtrain), c(nrow(train$data), ncol(train$data) + num_new_cat_features))
```

Reply:
Resolved via fab88db.